realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -605,7 +625,9 @@ public void onResult(int count) {
                 RealmMigration realmMigration = (migration == null) ? configuration.getMigration() : migration;
                 DynamicRealm realm = null;
                 try {
-                    realm = DynamicRealm.getInstance(configuration);
+                    // Create a DynamicRealm WITHOUT putting it into a RealmCache to avoid recursive locks and call init
+                    // steps multiple times (copy asset file / initialData transaction).
+                    realm = DynamicRealm.createInstance(configuration);
```

Comment:
`createInstance` is package protected, so user shouldn't call it to create Realm bypass cache. But if user use `Realm.getInstance()` in the migration block, the cache counter actually still works like before, so it is just the same like before.